src/nspawn/nspawn.c

```diff
@@ -2523,6 +2524,23 @@ static int on_orderly_shutdown(sd_event_source *s, const struct signalfd_siginfo
         return 0;
 }
 
+static int on_sigchld(sd_event_source *s, const struct signalfd_siginfo *si, void *userdata) {
+        pid_t pid;
+
+        // Because we set PR_SET_CHILD_REAPER, orphan processes may be
+        // reparented to us.
+        pid = PTR_TO_PID(userdata);
+        if (pid == si->ssi_pid || (arg_clone_ns_flags & CLONE_NEWPID)) {
```

Comment:
> I wasn't able to reproduce it, but I suspect that some (older) kernels incorrectly fill ssi_pid when the child is running in a private PID namespace.

Actually, I reproduced with `4.8.11-300.fc25.x86_64`.
I guess the problem was [`UINT32_TO_PTR(pid)`](https://github.com/systemd/systemd/pull/4855#discussion_r91688967) (instead of the `PID_TO_PTR(*pid)`)
```
3972            sd_event_add_signal(event, NULL, SIGCHLD, on_sigchld, UINT32_TO_PTR(pid));
(gdb) p *pid
$7 = 4246
(gdb) c
Continuing.

Breakpoint 1, on_sigchld (s=0x555555599c00, si=0x555555599c68, userdata=0x7fffffffe300) at src/nspawn/nspawn.c:2532
2532            pid = PTR_TO_UINT32(userdata);
2533            if (pid == si->ssi_pid) {
(gdb) p pid
$8 = -7424
(gdb) p si->ssi_pid
$9 = 4246
(gdb) q
```
I think we don't need `|| arg_clone_ns_flags & CLONE_NEWPID`